File: routers/v2/activity/activity.py
import hikari
import linkd
from fastapi import APIRouter, HTTPException, Depends, Query
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from typing import Optional
from datetime import datetime, timedelta
import coc

from utils.database import MongoClient, OldMongoClient
from utils.security import check_authentication
from utils.config import Config
from utils.custom_coc import CustomClashClient
from utils.sentry_utils import capture_endpoint_errors
from .activity_models import (
    GuildActivitySummary,
    ClanActivity,
    InactivePlayersResponse,
    InactivePlayer
)

config = Config()
security = HTTPBearer()

router = APIRouter(prefix="/v2/activity", tags=["Activity & Inactivity"], include_in_schema=True)


@router.get("/guild-summary",
            name="Get guild activity summary",
            response_model=GuildActivitySummary)
@linkd.ext.fastapi.inject
@check_authentication
@capture_endpoint_errors
async def get_guild_activity_summary(
    guild_id: int,
    inactive_threshold_days: int = 7,
    user_id: str = None,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    *,
    mongo: MongoClient,
    rest: hikari.RESTApp,
    coc_client: CustomClashClient
) -> GuildActivitySummary:
    """
    Get server-wide activity overview across all clans.

    Provides aggregate statistics including:
    - Total members and activity rates
    - Donation statistics
    - Clan-by-clan breakdown

    Args:
        guild_id: Discord server ID
        inactive_threshold_days: Days without activity to consider inactive (default 7)

    Returns:
        Comprehensive activity summary for the server
    """
    # Verify server exists and user has access
    server = await mongo.server_db.find_one({"server": guild_id})
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")

    # Get all clans for this server
    clans = await mongo.clan_db.find({"server": guild_id}).to_list(length=None)
    if not clans:
        return GuildActivitySummary(
            guild_id=guild_id,
            total_clans=0,
            total_members=0,
            total_active_members=0,
            total_inactive_members=0,
            overall_activity_rate=0.0,
            total_donations_sent=0,
            total_donations_received=0,
            clans=[]
        )

    clan_tags = [clan["tag"] for clan in clans]
    clan_name_map = {clan["tag"]: clan["name"] for clan in clans}

    # Calculate inactive threshold timestamp
    inactive_threshold = datetime.utcnow() - timedelta(days=inactive_threshold_days)

    # Aggregate statistics
    clan_activities = []
    total_members = 0
    total_active = 0
    total_inactive = 0
    total_donations_sent = 0
    total_donations_received = 0

    for clan_tag in clan_tags:
        try:
            # Fetch current clan data from CoC API
            clan = await coc_client.get_clan(clan_tag)

            # Get player activity data from database
            player_tags = [member.tag for member in clan.members]
            player_stats = await OldMongoClient.player_stats.find(
                {"tag": {"$in": player_tags}}
            ).to_list(length=None)

            # Create map of last seen timestamps
            last_seen_map = {}
            for player in player_stats:
                if "last_online" in player:
                    last_seen_map[player["tag"]] = player["last_online"]

            # Calculate clan statistics
            active_count = 0
            inactive_count = 0
            donations_sent = 0
            donations_received = 0
            total_trophies = 0
            attacks_wins = 0

            for member in clan.members:
                # Check if player is active
                last_seen = last_seen_map.get(member.tag)
                if last_seen and last_seen >= inactive_threshold:
                    active_count += 1
                else:
                    inactive_count += 1

                # Aggregate donations and stats
                donations_sent += member.donations
                donations_received += member.received
                total_trophies += member.trophies
                attacks_wins += member.attack_wins

            member_count = len(clan.members)
            activity_rate = (active_count / member_count * 100) if member_count > 0 else 0.0
            avg_donations_sent = donations_sent / member_count if member_count > 0 else 0.0
            avg_donations_received = donations_received / member_count if member_count > 0 else 0.0
            avg_trophies = total_trophies / member_count if member_count > 0 else 0.0

            clan_activities.append(ClanActivity(
                clan_tag=clan_tag,
                clan_name=clan_name_map.get(clan_tag, clan.name),
                total_members=member_count,
                active_members=active_count,
                inactive_members=inactive_count,
                activity_rate=activity_rate,
                average_donations_sent=avg_donations_sent,
                average_donations_received=avg_donations_received,
                total_donations_sent=donations_sent,
                total_donations_received=donations_received,
                total_attacks_wins=attacks_wins,
                average_trophies=avg_trophies
            ))

            # Update totals
            total_members += member_count
            total_active += active_count
            total_inactive += inactive_count
            total_donations_sent += donations_sent
            total_donations_received += donations_received

        except coc.NotFound:
            # Clan not found, skip
            continue
        except Exception as e:
            # Log error but continue with other clans
            print(f"Error fetching clan {clan_tag}: {e}")
            continue

    # Calculate overall activity rate
    overall_activity_rate = (total_active / total_members * 100) if total_members > 0 else 0.0

    return GuildActivitySummary(
        guild_id=guild_id,
        total_clans=len(clan_activities),
        total_members=total_members,
        total_active_members=total_active,
        total_inactive_members=total_inactive,
        overall_activity_rate=overall_activity_rate,
        total_donations_sent=total_donations_sent,
        total_donations_received=total_donations_received,
        clans=clan_activities
    )


@router.get("/inactive-players",
            name="Get inactive players",
            response_model=InactivePlayersResponse)
@linkd.ext.fastapi.inject
@check_authentication
@capture_endpoint_errors
async def get_inactive_players(
    guild_id: int,
    inactive_threshold_days: int = 7,
    min_townhall: Optional[int] = None,
    clan_tag: Optional[str] = None,
    limit: int = 100,
    offset: int = 0,
    user_id: str = None,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    *,
    mongo: MongoClient,
    rest: hikari.RESTApp,
    coc_client: CustomClashClient
) -> InactivePlayersResponse:
    """
    Get list of inactive members across server clans.

    Returns players who haven't been active within the threshold period.

    Args:
        guild_id: Discord server ID
        inactive_threshold_days: Days without activity to consider inactive (default 7)
        min_townhall: Optional minimum townhall level filter
        clan_tag: Optional specific clan to check (defaults to all server clans)
        limit: Maximum number of players to return (default 100)
        offset: Number of players to skip for pagination (default 0)

    Returns:
        List of inactive players with their details
    """
    # Verify server exists and user has access
    server = await mongo.server_db.find_one({"server": guild_id})
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")

    # Get clans to check
    if clan_tag:
        # Check specific clan
        clan_query = {"server": guild_id, "tag": clan_tag}
    else:
        # Check all server clans
        clan_query = {"server": guild_id}

    clans = await mongo.clan_db.find(clan_query).to_list(length=None)
    if not clans:
        return InactivePlayersResponse(
            guild_id=guild_id,
            inactive_threshold_days=inactive_threshold_days,
            players=[],
            total_count=0,
            limit=limit,
            offset=offset
        )

    clan_tags = [clan["tag"] for clan in clans]
    clan_name_map = {clan["tag"]: clan["name"] for clan in clans}

    # Calculate inactive threshold timestamp
    inactive_threshold = datetime.utcnow() - timedelta(days=inactive_threshold_days)

    # Collect inactive players
    all_inactive_players = []

    for clan_tag_check in clan_tags:
        try:
            # Fetch current clan data from CoC API
            clan = await coc_client.get_clan(clan_tag_check)

            # Get player activity data from database
            player_tags = [member.tag for member in clan.members]
            player_stats = await OldMongoClient.player_stats.find(
                {"tag": {"$in": player_tags}}
            ).to_list(length=None)

            # Create map of last seen timestamps
            last_seen_map = {
                player["tag"]: player.get("last_online")
                for player in player_stats
                if "last_online" in player
            }

            # Find inactive players
            for member in clan.members:
                # Apply townhall filter if specified
                if min_townhall and member.town_hall < min_townhall:
                    continue

                # Check if player is inactive
                last_seen = last_seen_map.get(member.tag)

                # Consider inactive if:
                # 1. No last_seen record (never tracked)
                # 2. Last seen is older than threshold
                is_inactive = (not last_seen) or (last_seen < inactive_threshold)

                if is_inactive:
                    # Calculate days inactive
                    if last_seen:
                        days_inactive = (datetime.utcnow() - last_seen).days
                    else:
                        days_inactive = None

                    all_inactive_players.append(InactivePlayer(
                        player_tag=member.tag,
                        player_name=member.name,
                        clan_tag=clan_tag_check,
                        clan_name=clan_name_map.get(clan_tag_check, clan.name),
                        townhall_level=member.town_hall,
                        role=member.role.name if member.role else "Member",
                        last_seen=last_seen,
                        days_inactive=days_inactive,
                        trophies=member.trophies,
                        donations_sent=member.donations,
                        donations_received=member.received
                    ))

        except coc.NotFound:
            # Clan not found, skip
            continue
        except Exception as e:
            # Log error but continue with other clans
            print(f"Error fetching clan {clan_tag_check}: {e}")
            continue

    # Sort by days inactive (most inactive first)
    all_inactive_players.sort(
        key=lambda p: p.days_inactive if p.days_inactive is not None else 9999,
        reverse=True
    )

    # Apply pagination
    total_count = len(all_inactive_players)
    paginated_players = all_inactive_players[offset:offset + limit]

    return InactivePlayersResponse(
        guild_id=guild_id,
        inactive_threshold_days=inactive_threshold_days,
        players=paginated_players,
        total_count=total_count,
        limit=limit,
        offset=offset
    )

File: internal/routes/v2/activity.go
package v2

import (
	"sort"
	"strconv"
	"time"

	modelsv2 "github.com/ClashKingInc/ClashKingAPI/internal/models/v2"
	apptypes "github.com/ClashKingInc/ClashKingAPI/internal/utils"
	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

// guildSummary godoc
// @Summary Get guild activity summary
// @Description Returns clan and member activity totals for a Discord guild.
// @Tags Activity & Inactivity
// @Produce json
// @Param guild_id query int true "Discord guild ID"
// @Param inactive_threshold_days query int false "Days without login to be considered inactive (default 7)"
// @Success 200 {object} map[string]interface{}
// @Failure 400 {object} map[string]interface{}
// @Failure 404 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /v2/guild-summary [get]
func guildSummary(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		guildID, _ := strconv.ParseInt(c.Query("guild_id"), 10, 64)
		inactiveDays := activityParseIntDefault(c.Query("inactive_threshold_days"), 7)
		if guildID == 0 {
			return apptypes.Error(fiber.StatusBadRequest, "guild_id is required")
		}

		var serverDoc bson.M
		if err := a.Store.C.ServerDB.FindOne(c.UserContext(), bson.M{"server": guildID}).Decode(&serverDoc); err != nil {
			if err == mongo.ErrNoDocuments {
				return apptypes.Error(fiber.StatusNotFound, "Server not found")
			}
			return err
		}

		clans, err := serverClans(c, a, guildID)
		if err != nil {
			return err
		}
		if len(clans) == 0 {
			return apptypes.JSON(c, fiber.StatusOK, modelsv2.GuildSummaryResponse{
				GuildID: guildID, Clans: []modelsv2.GuildSummaryClanRow{},
			})
		}

		cutoffUnix := time.Now().Unix() - int64(inactiveDays)*86400

		type memberEntry struct {
			tag       string
			trophies  int
			donations int
			received  int
		}
		type clanEntry struct {
			tag     string
			name    string
			members []memberEntry
		}

		clanEntries := make([]clanEntry, 0, len(clans))
		allTags := make([]string, 0)
		for _, clan := range clans {
			tag := activityAsString(clan["tag"])
			name := activityAsString(clan["name"])
			cocClan, err := a.Clash.GetClan(c.UserContext(), tag)
			if err != nil || cocClan == nil {
				continue
			}
			members := make([]memberEntry, 0, len(cocClan.Members))
			for _, m := range cocClan.Members {
				members = append(members, memberEntry{
					tag: m.Tag, trophies: m.Trophies,
					donations: m.Donations, received: m.Received,
				})
				allTags = append(allTags, m.Tag)
			}
			clanEntries = append(clanEntries, clanEntry{tag: tag, name: name, members: members})
		}

		// Batch-query last_online from player_stats.
		lastOnlineMap := make(map[string]int64)
		if len(allTags) > 0 {
			loCur, err := a.Store.C.PlayerStats.Find(
				c.UserContext(),
				bson.M{"tag": bson.M{"$in": allTags}},
				options.Find().SetProjection(bson.M{"tag": 1, "last_online": 1, "_id": 0}),
			)
			if err == nil {
				var loDocs []bson.M
				if err := loCur.All(c.UserContext(), &loDocs); err == nil {
					for _, doc := range loDocs {
						t, _ := doc["tag"].(string)
						lastOnlineMap[t] = activityAsInt64(doc["last_online"])
					}
				}
			}
		}

		summaries := make([]modelsv2.GuildSummaryClanRow, 0, len(clanEntries))
		totalMembers, totalActive, totalInactive := 0, 0, 0
		totalDonSent, totalDonRcvd := 0, 0

		for _, ce := range clanEntries {
			active, inactive := 0, 0
			donSent, donRcvd, trophySum := 0, 0, 0
			for _, m := range ce.members {
				if lastOnlineMap[m.tag] > cutoffUnix {
					active++
				} else {
					inactive++
				}
				donSent += m.donations
				donRcvd += m.received
				trophySum += m.trophies
			}
			n := len(ce.members)
			rate, avgDonSent, avgDonRcvd, avgTrophies := 0.0, 0.0, 0.0, 0.0
			if n > 0 {
				rate = float64(active) / float64(n) * 100
				avgDonSent = float64(donSent) / float64(n)
				avgDonRcvd = float64(donRcvd) / float64(n)
				avgTrophies = float64(trophySum) / float64(n)
			}
			summaries = append(summaries, modelsv2.GuildSummaryClanRow{
				ClanTag:                  ce.tag,
				ClanName:                 ce.name,
				TotalMembers:             n,
				ActiveMembers:            active,
				InactiveMembers:          inactive,
				ActivityRate:             rate,
				AverageDonationsSent:     avgDonSent,
				AverageDonationsReceived: avgDonRcvd,
				TotalDonationsSent:       donSent,
				TotalDonationsReceived:   donRcvd,
				AverageTrophies:          avgTrophies,
			})
			totalMembers += n
			totalActive += active
			totalInactive += inactive
			totalDonSent += donSent
			totalDonRcvd += donRcvd
		}

		overallRate := 0.0
		if totalMembers > 0 {
			overallRate = float64(totalActive) / float64(totalMembers) * 100
		}
		return apptypes.JSON(c, fiber.StatusOK, modelsv2.GuildSummaryResponse{
			GuildID:                guildID,
			TotalClans:             len(summaries),
			TotalMembers:           totalMembers,
			TotalActiveMembers:     totalActive,
			TotalInactiveMembers:   totalInactive,
			OverallActivityRate:    overallRate,
			TotalDonationsSent:     totalDonSent,
			TotalDonationsReceived: totalDonRcvd,
			Clans:                  summaries,
		})
	}
}

// inactivePlayers godoc
// @Summary Get inactive players
// @Description Returns players in the guild that are considered inactive by the provided threshold.
// @Tags Activity & Inactivity
// @Produce json
// @Param guild_id query int true "Discord guild ID"
// @Param inactive_threshold_days query int false "Days before a player is considered inactive"
// @Param min_townhall query int false "Minimum town hall level"
// @Param clan_tag query string false "Restrict results to a single clan"
// @Param limit query int false "Maximum number of results"
// @Param offset query int false "Number of results to skip"
// @Success 200 {object} map[string]interface{}
// @Failure 400 {object} map[string]interface{}
// @Failure 404 {object} map[string]interface{}
// @Failure 500 {object} map[string]interface{}
// @Router /v2/inactive-players [get]
func inactivePlayers(a apptypes.Deps) fiber.Handler {
	return func(c *fiber.Ctx) error {
		guildID, _ := strconv.ParseInt(c.Query("guild_id"), 10, 64)
		if guildID == 0 {
			return apptypes.Error(fiber.StatusBadRequest, "guild_id is required")
		}
		inactiveDays := activityParseIntDefault(c.Query("inactive_threshold_days"), 7)
		minTownHall := activityParseIntDefault(c.Query("min_townhall"), 0)
		limit := activityParseIntDefault(c.Query("limit"), 100)
		offset := activityParseIntDefault(c.Query("offset"), 0)
		clanFilter := c.Query("clan_tag")

		var serverDoc bson.M
		if err := a.Store.C.ServerDB.FindOne(c.UserContext(), bson.M{"server": guildID}).Decode(&serverDoc); err != nil {
			if err == mongo.ErrNoDocuments {
				return apptypes.Error(fiber.StatusNotFound, "Server not found")
			}
			return err
		}

		clans, err := serverClans(c, a, guildID)
		if err != nil {
			return err
		}

		cutoffUnix := time.Now().Unix() - int64(inactiveDays)*86400
		nowUnix := time.Now().Unix()

		type memberInfo struct {
			tag      string
			name     string
			clanTag  string
			clanName string
			role     string
			trophies int
			donSent  int
			donRcvd  int
		}

		allMembers := make([]memberInfo, 0)
		for _, clan := range clans {
			tag := activityAsString(clan["tag"])
			if clanFilter != "" && tag != clanFilter {
				continue
			}
			name := activityAsString(clan["name"])
			cocClan, err := a.Clash.GetClan(c.UserContext(), tag)
			if err != nil || cocClan == nil {
				continue
			}
			for _, m := range cocClan.Members {
				allMembers = append(allMembers, memberInfo{
					tag:      m.Tag,
					name:     m.Name,
					clanTag:  tag,
					clanName: name,
					role:     string(m.Role),
					trophies: m.Trophies,
					donSent:  m.Donations,
					donRcvd:  m.Received,
				})
			}
		}

		// Batch-query player_stats for last_online and townhall.
		type playerStatsData struct {
			lastOnline int64
			townhall   int
		}
		playerDataMap := make(map[string]playerStatsData)
		if len(allMembers) > 0 {
			allTags := make([]string, len(allMembers))
			for i, m := range allMembers {
				allTags[i] = m.tag
			}
			loCur, err := a.Store.C.PlayerStats.Find(
				c.UserContext(),
				bson.M{"tag": bson.M{"$in": allTags}},
				options.Find().SetProjection(bson.M{"tag": 1, "last_online": 1, "townhall": 1, "_id": 0}),
			)
			if err == nil {
				var loDocs []bson.M
				if err := loCur.All(c.UserContext(), &loDocs); err == nil {
					for _, doc := range loDocs {
						t, _ := doc["tag"].(string)
						playerDataMap[t] = playerStatsData{
							lastOnline: activityAsInt64(doc["last_online"]),
							townhall:   activityAsInt(doc["townhall"]),
						}
					}
				}
			}
		}

		items := make([]modelsv2.InactivePlayerItem, 0)
		for _, m := range allMembers {
			pd := playerDataMap[m.tag]

			// Apply min_townhall filter using player_stats townhall (skip if townhall unknown).
			if minTownHall > 0 && pd.townhall > 0 && pd.townhall < minTownHall {
				continue
			}
			// Only include players who have not been online since the cutoff.
			if pd.lastOnline > cutoffUnix {
				continue
			}

			var daysInactivePtr *int
			if pd.lastOnline > 0 {
				days := int((nowUnix - pd.lastOnline) / 86400)
				daysInactivePtr = &days
			}
			var thPtr *int
			if pd.townhall > 0 {
				th := pd.townhall
				thPtr = &th
			}
			items = append(items, modelsv2.InactivePlayerItem{
				Tag:               m.tag,
				Name:              m.name,
				ClanTag:           m.clanTag,
				ClanName:          m.clanName,
				Role:              m.role,
				Trophies:          m.trophies,
				Townhall:          thPtr,
				DaysInactive:      daysInactivePtr,
				DonationsSent:     m.donSent,
				DonationsReceived: m.donRcvd,
			})
		}

		// Sort most-inactive first; players with no last_online data go last.
		sort.Slice(items, func(i, j int) bool {
			di, dj := -1, -1
			if items[i].DaysInactive != nil {
				di = *items[i].DaysInactive
			}
			if items[j].DaysInactive != nil {
				dj = *items[j].DaysInactive
			}
			return di > dj
		})

		if offset > len(items) {
			offset = len(items)
		}
		end := offset + limit
		if end > len(items) {
			end = len(items)
		}
		return apptypes.JSON(c, fiber.StatusOK, modelsv2.InactivePlayersResponse{
			GuildID:               guildID,
			InactiveThresholdDays: inactiveDays,
			Players:               items[offset:end],
			TotalCount:            len(items),
			Limit:                 limit,
			Offset:                offset,
		})
	}
}

func serverClans(c *fiber.Ctx, a apptypes.Deps, guildID int64) ([]bson.M, error) {
	cur, err := a.Store.C.ClanDB.Find(c.UserContext(), bson.M{"server": guildID})
	if err != nil {
		return nil, err
	}
	var clans []bson.M
	if err := cur.All(c.UserContext(), &clans); err != nil {
		return nil, err
	}
	return clans, nil
}

func activityAsString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func activityAsInt(v any) int {
	switch x := v.(type) {
	case int32:
		return int(x)
	case int64:
		return int(x)
	case int:
		return x
	case float64:
		return int(x)
	}
	return 0
}

func activityAsInt64(v any) int64 {
	switch x := v.(type) {
	case int32:
		return int64(x)
	case int64:
		return x
	case int:
		return int64(x)
	case float64:
		return int64(x)
	}
	return 0
}

func activityParseIntDefault(raw string, fallback int) int {
	if raw == "" {
		return fallback
	}
	value, err := strconv.Atoi(raw)
	if err != nil {
		return fallback
	}
	return value
}


